### ISSUE TITLE
Why padding grids instead of margin ones?

### ISSUE THREAD

**Issue by corysimmons** [twbs/bootstrap#19026]
Like, isn't this the same thing (negative margins on rows) except it has real gutter in between elements? http://codepen.io/corysimmons/pen/MKVmqM?editors=1100

Margin based lets the background colors shine through.

Just curious if I'm missing something here... :dancer:

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/MKVmqM.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 10, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 12, column 9: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 13, column 9: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 14, column 9: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 15, column 9: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 16, column 9: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 19, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 20, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 21, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 22, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 23, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 24, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 25, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 26, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 27, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 28, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 29, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 30, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s
* line 31, column 5: [E013](https://github.com/twbs/bootlint/wiki/E013): Only columns (`.col-*-*`) may be children of `.row`s

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by corysimmons** [comment]
Updated it even though I'm not using Bootstrap in the example.

**Comment by cvrebert** [comment]
I think #16358 answers this, albeit `calc()` wasn't an option at that point since we still supported IE8.

**Comment by corysimmons** [comment]
http://v4-alpha.getbootstrap.com/getting-started/browsers-devices/#supporting-internet-explorer-8

Now that you're not supporting IE8 do you have plans on making a new grid?

**Comment by cvrebert** [comment]
> do you have plans on making a new grid?

Not presently, although it's not presently out of the question either. The decision would be up to @mdo.

(Technically, there's already another new grid in v4: When the optional support for Flexbox (http://v4-alpha.getbootstrap.com/getting-started/flexbox/ ) is enabled, the grid switches to using Flexbox instead of floats.)

**Comment by corysimmons** [comment]
Alrighty. Just curious if there was a good reason for them. Thanks. :^)

**Comment by cvrebert** [comment]
Seems like a reasonable proposal to at least consider, now that `calc()` is in our arsenal.

**Comment by corysimmons** [comment]
Well let me know if you guys decide to upgrade and if you're interested in help.

**Comment by corysimmons** [comment]
In fact, here it is: http://codepen.io/corysimmons/pen/dGmEBW?editors=1100

Let me know if you want a PR.

**Comment by danielgamble** [comment]
I think the @mdo comment in #16358 might be a little redundant now we have access to `calc()`. With Cory on this and thing the `margin` grid far outshines `padding` as it means we don't need to separate layout divs from styling divs as much. Although i still wouldn't really advise adding extra classes to `col-*` elements to add style it definitely opens up that avenue :)

**Comment by corysimmons** [comment]
### Caveats:
- `calc` doesn't work in IE8. That's 1.27% of users you're not supporting anyway, but even so calc-polyfill works well in IE8.
- `calc` doesn't work in Android Browser 4.3 and below. That's 1.63% of users. I'm not sure if you're supporting them anyway, but calc-polyfill also fixes this.

Stats from http://caniuse.com/#feat=calc

### Pros
- Users don't have to put an element immediately inside of a column in order to see the background

The biggest complaint I've heard about Bootstrap is you generate a ton of markup.

Not only do you have to nest row elements for your grid:

```html
<div class="row">
  <div class="column">
    <div class="row">
      <div class="column">
        1a
      </div>
      <div class="column">
        2a
      </div>
    </div>
  </div>
  <div class="column">
    1
  </div>
</div>
```

But if you want the background of a row shine through, you have to nest even more elements inside the columns:

```html
<div class="row">
  <div class="column">
    <div class="row">
      <div class="column">
        <div>
          1a
        </div>
      </div>
      <div class="column">
        <div>
          2a
        </div>
      </div>
    </div>
  </div>
  <div class="column">
    <div>
      1
    </div>
  </div>
</div>
```

Here's the exact same layout using real gutters:

```html
<div class="row">
  <div class="col-sm-6">
    <div class="row">
      <div class="col-sm-6">
        1a
      </div>
      <div class="col-sm-6">
        2a
      </div>
    </div>
  </div>
  <div class="col-sm-6">
    1
  </div>
</div>
```

https://youtu.be/ueZ6tvqhk8U?t=17

**Comment by cvrebert** [comment]
v4 only officially supports Android>=5

**Comment by corysimmons** [comment]
Ah crap.

Just found a fairly important caveat. When you apply a bg color to the row it isn't aligned flush with the items in that row.

I'm not sure if you guys can think of some cool way to beat this as you're more familiar with these negative margin type grids.

I came up with this hacky solution (that I'm sure no one is gonna go for): http://codepen.io/corysimmons/pen/EPLYaM?editors=1100

It occupies `.row:before` and requires the user to specify their `.row` background color (or border) on `.row:before`.

Maybe you guys could consider switching to a nth-grid? I made this the other night: http://codepen.io/corysimmons/pen/XXERVr?editors=1100

**Comment by corysimmons** [comment]
Actually there's another glaring caveat with my solution. You can't add bg to these rows. They won't align flush with the columns inside.

You can add bg to the container and it works (as in my first example), or you can even do some gross hackery stuff to apply bg to the row's `:before` (but that's not good).

This might be why @mdo settled on padding grids.

**Comment by mdo** [comment]
> Actually there's another glaring caveat with my solution. You can't add bg to these rows. They won't align flush with the columns inside.
>
> This might be why @mdo settled on padding grids.

Indeed it is, but for the columns mostly (and the rows if you need it for all columns).

---

[My comment here](https://github.com/twbs/bootstrap/issues/16358#issuecomment-96013008) sums it up nicely—I'd rather say `25%` and get exactly `25%` instead of having to do extra math or rely on `calc` unnecessarily. To me `calc` is a last resort with this stuff. This also fits with the `border-box` model we're using throughout Bootstrap.

We won't be changing this in v4. I want to keep the grid rather familiar in functionality and behavior when folks move from v3 to v4.

**Comment by mmjaeger** [comment]
how would this work using flex box?

**Comment by corysimmons** [comment]
@mdo The `border-box` model you're using produces a lot of extra markup. Why not replace it if it's possible and make Bootstrap that much cleaner/sensible?

I can accept that it's too late for 4 and it's going to have a padding grid, but you do realize this is a problem right? Why not leave this open and let the community figure out how to solve it?

**Comment by mdo** [comment]
> The border-box model you're using produces a lot of extra markup.

I do realize it can lead to more markup in the specific instance you're referring to, like so: https://jsbin.com/fifejagepi/edit?html,output. I don't think that's the majority use case though.

> Why not replace it if it's possible and make Bootstrap that much cleaner/sensible?

It'd be more sensible for your use case, and I understand why you'd want it to be the default, but this is just as easy to understand for the masses :).

> I can accept that it's too late for 4 and it's going to have a padding grid, but you do realize this is a problem right?

I do realize there's room for improvement, but I'm not out to solve every problem. We do what we can for the largest amount of people possible without adding unnecessary complexity.

> Why not leave this open and let the community figure out how to solve it?

That doesn't really work out too well. If you want margin, write the custom CSS to do it and open source it for folks to drop in and replace the default grid system. That's how folks did it when they wanted flexbox ahead of us offering any official support for it.

**Comment by twbs-lmvtfy** [comment]
Hi @mdo!

You appear to have posted a live example (https://jsbin.com/fifejagepi/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by corysimmons** [comment]
:+1:

**Comment by corysimmons** [comment]
Actually, :-1: 

Padding grids suck (result in more markup) and I want this issue to serve as a record that I tried and don't agree with you.

![image](https://cloud.githubusercontent.com/assets/224636/12924065/4a749894-cf26-11e5-8d1f-4ad1dfd4d3b3.png)

http://codepen.io/corysimmons/pen/MKzPWW?editors=1100 solves this in spades except you have to wrap a row in an element with bg color (or use a clever `&:before` hack) if you *happen* to want bg on it (vs. having to nest an element inside *every single element* in your row).

Get a 90k starred repo and sleep on the most important part of it because you think padding based grids are less confusing than margin based ones.

Congrats on tricking people into thinking v4 is better by introducing flexbox.

![image](https://cloud.githubusercontent.com/assets/224636/12924933/d59db92a-cf29-11e5-9720-37194b7e68ba.png)

Yeah yeah, I know I'm a poopy head and "not being very constructive".

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/MKzPWW.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 8, column 1: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`
* line 52, column 3: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by seigler** [comment]
background-clip is well supported -- you can use either padding- or margin-based grids without issues if you no longer care about IE8.

**Comment by corysimmons** [comment]
@seigler How would that work? Tried doing stuff with it to clip the row bg on [this fork](http://codepen.io/corysimmons/pen/BjGvQw?editors=1100) to no avail.

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/BjGvQw.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 8, column 1: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`
* line 52, column 1: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by seigler** [comment]
@corysimmons I think I misunderstood the issue; I see the problem more clearly now. Thanks for the example.

**Comment by corysimmons** [comment]
Sorry if I was unclear. Here are some reduced cases to illustrate the problem:

- [Bootstrap padding grids requiring 2 divs to give bg to .row](http://codepen.io/corysimmons/pen/wMQRjj?editors=1100) (which isn't even flush, and if they have more than 2 columns they will have that many more superfluous inner divs)
- [proposed margin grid with :before trick](http://codepen.io/corysimmons/pen/yeQGjw?editors=1100) (no extra elements for any number of columns, flush)
- [proposed margin grid with row wrapped in any bg element](http://codepen.io/corysimmons/pen/gPQZjX?editors=1100) (1 wrapping element that doesn't hurt row alignment, any number of columns, flush)

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/wMQRjj.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 8, column 1: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/gPQZjX.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 9, column 3: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @corysimmons!

You appear to have posted a live example (http://codepen.io/corysimmons/pen/yeQGjw.html), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 8, column 1: [E003](https://github.com/twbs/bootlint/wiki/E003): Found one or more `.row`s that were not children of a grid column or descendants of a `.container` or `.container-fluid`

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by corysimmons** [comment]
Can you add an option to that bot that if I put `NO-BOT` in my comment it won't comment on that particular message? Not every codepen is a bootstrap codepen.

**Comment by corysimmons** [comment]
https://github.com/corysimmons/bootstrap-margin-grid WIP, if it gets attention I'll make it a drop-in replacement.

**Comment by corysimmons** [comment]
Chiming in for posterity's sake:

Padding grids are pretty sweet because they push component architecture (i.e. being able to dump a block of markup in pretty much anywhere and know it's going to look right).

There _is_ a reason for margin grids: much terser markup.

Once newbs graduate to component stuff (React, Vue, etc.), you'll see the benefits of this.

I was a jerk because I wasn't satisfied with your answer @mdo 

It looked like you were saying "math is hard" and I thought that was a pretty messed up reason for making tens-of-thousands of devs jobs a bit harder, but there is definitely a huge argument to be made for padding grids.

fwiw, I had thought of that at the time, but it wasn't until I started building really component-heavy stuff that I noticed the importance of this.

All this to say, I'm sorry for being a jerk here. Thanks for all your work throughout the years.

---

Other relevant rules for reusable components: http://rscss.io/layouts.html#avoid-positioning-properties
